pkg/agent/watcher.go

```diff
@@ -84,78 +86,47 @@ func (w *Watcher) Start() {
 			}
 		}
 	}()
-	err = watcher.Add(w.ConfigDir)
-	if err != nil {
-		log.Fatal(err)
-	}
-	log.Println("Watching", w.ConfigDir)
+	log.Println("Watching", w.configDir)
 	<-done
 }
 
-func (w *Watcher) ParseConfig(modelConfigs modelconfig.ModelConfigs) {
-	timeNow := time.Now()
+func (w *Watcher) parseConfig(modelConfigs modelconfig.ModelConfigs) {
 	for _, modelConfig := range modelConfigs {
-		modelName := modelConfig.Name
-		modelSpec := modelConfig.Spec
-		log.Println("Name:", modelName, "Spec:", modelSpec)
-		oldModel, ok := w.ModelTracker[modelName]
-		if !ok {
-			w.ModelTracker[modelName] = ModelWrapper{
-				ModelSpec:  &modelSpec,
-				Time:       timeNow,
-				Stale:      true,
-				Redownload: true,
-			}
-		} else {
-			isStale := true
-			reDownload := true
-			if oldModel.ModelSpec != nil {
-				isStale = !cmp.Equal(*oldModel.ModelSpec, modelSpec)
-				reDownload = !cmp.Equal(oldModel.ModelSpec.StorageURI, modelSpec.StorageURI)
-				log.Println("same", !isStale, *oldModel.ModelSpec, modelSpec)
-			}
-			// Need to store new time, TODO: maybe worth to have seperate map?
-			w.ModelTracker[modelName] = ModelWrapper{
-				ModelSpec:  &modelSpec,
-				Time:       timeNow,
-				Stale:      isStale,
-				Redownload: reDownload,
-			}
+		name, spec := modelConfig.Name, modelConfig.Spec
+		existing, exists := w.modelTracker[name]
+		if !exists {
+			// New - add
+			w.modelTracker[name] = modelWrapper{Spec: &spec}
+			w.modelAdded(name, &spec)
+		} else if !cmp.Equal(spec, *existing.Spec) {
+			existing.Spec, existing.stale = &spec, false
+			// Changed - replace
+			w.modelRemoved(name)
+			w.modelAdded(name, &spec)
 		}
 	}
-	// TODO: Maybe make parallel and more efficient?
-	for modelName, modelWrapper := range w.ModelTracker {
-		if modelWrapper.Time.Before(timeNow) {
-			delete(w.ModelTracker, modelName)
-			channel, ok := w.Puller.ChannelMap[modelName]
-			if !ok {
-				log.Println("Model", modelName, "was never added to channel map")
-			} else {
-				event := EventWrapper{
-					ModelName:      modelName,
-					ModelSpec:      nil,
-					LoadState:      ShouldUnload,
-					ShouldDownload: false,
-				}
-				log.Println("Sending event", event)
-				channel.EventChannel <- event
-			}
+	for name, wrapper := range w.modelTracker {
+		if wrapper.stale {
```

Comment:
I converted the modelconfig to a map and use that to find the removed ones, I think this works too and slightly more efficient but a bit hard to follow the logic.

Reply:
I was aiming to minimze iterations over the map/slices (i.e. 1 pass over each). Agree the way that the stale flag is toggled could be a bit confusing to read... let me add a couple more comments and if it's still not clear then could change to use an intermediate map